documentation/architecture.md

```diff
@@ -0,0 +1,337 @@
+The command is sent to `cc-proxy` who forwards it to the right hyperstart instance running
+in the appropriate guest.
+4. Spawn the `cc-shim` process for it to forward the output streams (stderr and stdout) and the `exec`
+command exit code to Docker.
+
+Now the `exec`'ed process is running in the virtual machine, sharing the UTS, PID, mount and IPC
+namespaces with the container's init process.
+
+##### `kill`
+
+When sending the OCI `kill` command, container runtimes should send a [UNIX signal](https://en.wikipedia.org/wiki/Unix_signal)
+to the container process.
+In the Clear Containers context, this means `cc-oci-runtime` needs a way to send a signal
+to the container process within the virtual machine. As `cc-shim` is responsible for
+forwarding signals to its associated running containers, `cc-oci-runtime` naturally
+calls `kill` on the `cc-shim` PID.
+
+However, `cc-shim` is not able to trap `SIGKILL` and `SIGSTOP` and thus `cc-oci-runtime`
+needs to follow a different code path for those 2 signals.
+Instead of `kill`'ing the `cc-shim` PID, it will go through the following steps:
+
+1. `cc-oci-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it is trying to `kill` is running.
+2. `cc-oci-runtime` sends an hyperstart `KILLCONTAINER` command to `kill` the container running
+on the guest. The command is sent to `cc-proxy` who forwards it to the right hyperstart instance
+running in the appropriate guest.
+
+##### `delete`
+
+`docker delete` is about deleting all resources held by a stopped/killed container. Running
+containers can not be `delete`d unless the OCI runtime is explictly being asked to. In that
+case it will first `kill` the container and only then `delete` it.
+
+The resources held by a Clear Container are quite different from the ones held by a host
+namespace container e.g. run by `runc`. `cc-oci-runtime` needs mostly to delete the pod
+holding the stopped container on the virtual machine, shut the hypervisor down and finally
+delete all related proxy resources:
+
+1. `cc-oci-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it is trying to to `delete` is running.
+2. `cc-oci-runtime` sends an hyperstart `DESTROYPOD` command to `destroy` the pod holding the
+container running on the guest. The command is sent to `cc-proxy` who forwards it to the right
+hyperstart instance running in the appropriate guest.
+3. After deleting the last running pod, `hyperstart` will gracefully shut the virtual machine
+down.
+4. `cc-oci-runtime` sends the `BYE` command to `cc-proxy`, to let it know that a given virtual
+machine is shut down. `cc-proxy` will then clean all its internal resources associated with this
+VM.
+
+#### Proxy
+
+`cc-proxy` is a daemon offering access to the [`hyperstart`](https://github.com/hyperhq/hyperstart)
+VM agent to multiple clients like for example `cc-shim` and `cc-proxy`.
```

Comment:
This sentence defines `cc-proxy` by referring to `cc-proxy`. I think the following change captures the minimal definition:

"VM agent to multiple `cc-shim` clients."